### DISCUSSION TITLE
lack of specialization on scala 3

### DISCUSSION THREAD

**Discussion by pityka** [pityka/saddle#376] [Q&A] [answered]
@tnielens do you have any idea of a workaround?
What are your thoughts on some sort of manual code generation (templates)? Ugly but at least no more ambiguity on boxing.

**Comment by tnielens** [comment, marked as answer]
I'm not so keen on trying to roll out a custom solution for this. I fear this is a huge endeavor with little chance for a positive pros/cons balance. Iirc, from talks and scala github issues, `@specialized` introduced lots of complexity in scala2 and the language contributors didn't look back positively on the feature, hence no support in scala3.

I would rather hope for some future support in scala 3: https://contributors.scala-lang.org/t/status-of-specialization-in-scala-3/4628
Or wait for Valhalla inline types to land: https://openjdk.java.net/projects/valhalla/

In the meantime, saddle could still publish scala3 binaries, with a warning for performance characteristics in the doc. Like spire did: https://github.com/typelevel/spire/releases/tag/v0.18.0-M1

**Reply by pityka** [reply]
Yes I've seen what spire did. It is good for compatibility, early preview and maybe for scala.js .
